### DISCUSSION TITLE
Added: LDM Notes for April 7th, 2021

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#4650] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-04-07.md

## Agenda

- MVP Session

**Comment by jnm2** [comment]
> We revisited why we are looking at this feature as C# syntax, and not as a project file switch.

It could be a project file switch too, if the SDK started emitting global usings along with assembly attributes, or if you use a community NuGet package that does the same thing.

**Reply by HaloFour** [reply]
It's always been a project switch in VB.NET (via command line args) so there's precedent and an audience of developers that can relay their experiences with it.  Personally I liked project-wide imports in VB.NET and didn't find them confusing at all.

**Reply by alrz** [reply]
you could possibly make it like a project switch by adding a source file via msbuild the same way assembly attributes are added.

**Reply by jnm2** [reply]
(That's what I meant by emitting global usings along with assembly attributes.)

**Reply by alrz** [reply]
Was replying to HaloFour, I mean at that point it practically make no difference. Even without going through msbuild tricks to insert source, you can just link a file in directory build props.

**Comment by BreyerW** [comment]
> The second point is in source replacement. We are very unsure about this feature: in particular, the IDE experience isn't great. We've already received occasional bugs from users about IDE experiences with existing code-rewriting solutions such as Fody (such as debuggers appearing to misbehave), and we have additional concerns about the perf implications of such generators.

Just curious was ever a subset of that considered? That is additions only but extended to everything not just partials. This would enable for example automatic code instrumentation for testing or raising events in more complex scenarios than just INPC.

**Reply by HaloFour** [reply]
https://github.com/dotnet/csharplang/discussions/3748

I think there was a proposal for allowing only one part of a partial type to be marked as partial, with source generators specifically in mind.  The net result would be the same as if partial was completely optional.

**Reply by BreyerW** [reply]
@HaloFour  i believe this does NOT cover everything im talking about. Eg adding a call to diagnostic methods on start of existing method and its end (either right before all returns or before last `}` if existing method is `void`

**Reply by HaloFour** [reply]
It's considered modifications if it involved changing the body of existing code, even if what you were doing was "adding" to the beginning or end of the method.  That was really the only scenario that would have been supported by `original`/`replace`: https://github.com/dotnet/csharplang/issues/107

**Reply by HaloFour** [reply]
I completely agree, though.  AOP scenarios would be an amazing killer feature for source generators.

**Comment by orthoxerox** [comment]
Sad to see expression trees stuck in limbo. They're blocking language support for Roslyn syntax trees by their mere existence and are popular enough to be sustained indefinitely.

**Reply by MgSam** [reply]
Forgive my ignorance, but how do they block Roslyn syntax trees?

**Reply by orthoxerox** [reply]
By existing. If they didn't exist, everything would be using Roslyn syntax trees instead, and there would be demand for better support at the language level. But now everything except source generators already uses expressions and migrating away from them is a daunting task.

**Reply by CyrusNajmabadi** [reply]
> By existing. If they didn't exist, everything would be using Roslyn syntax trees instead

I really don't think that's the case.  For one thing, Roslyn syntax trees are C# and VB specific.  We wouldn't expect the abstract expression language to be concrete C# syntax trees.  Furthermore, our syntax trees are are jsut that... syntax.  Expression trees are AST that contain important semantic information in them.

**Comment by MgSam** [comment]
Would a versioned API work for adding features to Expressions? As-in, by default you get the current features (since versioning never existed prior). To get a new feature set, you explicitly opt-in to that version?

**Reply by jnm2** [reply]
@333fred I think this was mentioned briefly in the meeting.

**Reply by 333fred** [reply]
There is potential, but cross-assembly dependencies are a problem. What happens if assembly A generates trees v1, and B consumes v1, then A is updated at runtime now produces v2?

**Comment by huoyaoyuan** [comment]
> Cost-free delegates and LINQ

I wonder if any sort of auto vectorization could be possible. That may be closer to area of theoretical PL though.

**Comment by huoyaoyuan** [comment]
> Mixed-language projects

As C# evolving, mixing C# with VB shows less value since VB is almost stopped.

Mixing C# and F# aren't easy to use because their programming models are too different.

Mixing C# and IL may be useful. I'm interested in exposing things valid in IL but not allowed in C#.

Generally, I'm treating mixing C# with VB a proof of concept and architecture, that "we have a compiler model can handle two languages simultaneously".

**Reply by HaloFour** [reply]
@huoyaoyuan 

> As C# evolving, mixing C# with VB shows less value since VB is almost stopped.

The one place where I can see such mixing being valuable is when migrating a larger VB project over to C# iteratively.  Given the evolution of VB has effectively stopped that is something that I imagine organizations would be interested in.

**Reply by AartBluestoke** [reply]
and you could always migrate to c#8 under a current runtime, then roll forwards, even if in the future we complicate the migration. Microsoft is happy for such migration paths in many places, going from version A to G requires intermediate steps of C and E.

**Comment by SimonCropp** [comment]
> The second point is in source replacement. We are very unsure about this feature: in particular, the IDE experience isn't great. We've already received occasional bugs from users about IDE experiences with existing code-rewriting solutions such as Fody (such as debuggers appearing to misbehave), and we have additional concerns about the perf implications of such generators.

A few points

i would avoid referring to Fody as "source replacement" since that incorrectly describes what it does, and would lead to have an incorrect understanding

> We've already received occasional bugs from users about IDE experiences

 * how often? 
 * can u link to them? 
 * do you reach out to the fody team when they occur? 
 * when u fix/work around the bugs does that knowledge get pushed back into the fody codebase?
 
> we have additional concerns about the perf implications of such generators

 * runtime, or build time, or both?
 * do you have any numbers?
